### ISSUE TITLE
different performance between 23.03 and 23.04

### ISSUE THREAD

**Issue by cceyda** [triton-inference-server/server#5817]
**Description**

I'm seeing wildly different performance between 23.03 -> 23.04 (worse)
I'm using the same `model.onnx` file (also tried with file generated inside the trt but didn't make difference) while converting to `model.plan` & the same command for the conversion & the same `config.pbtxt` while serving the model. The only difference is the tritonserver & the equivalent tensorrt container used to generate the `model.plan`:

nvcr.io/nvidia/tritonserver:23.04-py3 & nvcr.io/nvidia/tensorrt:23.04-py3
nvcr.io/nvidia/tritonserver:23.03-py3 & nvcr.io/nvidia/tensorrt:23.03-py3

The command for conversion:
```
max_batch="64"
opt_shape="64"
exec_string=f"trtexec --onnx=models/my_model.onnx --saveEngine=my_model{max_batch}.plan --shapes=input_ids:{max_batch}x768,attention_mask:{max_batch}x768x768,position_ids:{max_batch}x768 --minShapes=input_ids:1x768,attention_mask:1x768x768,position_ids:1x768 --optShapes=input_ids:{opt_shape}x768,attention_mask:{opt_shape}x768x768,position_ids:{opt_shape}x768 --maxShapes=input_ids:{max_batch}x768,attention_mask:{max_batch}x768x768,position_ids:{max_batch}x768 --useCudaGraph --threads " 
!CUDA_MODULE_LOADING=LAZY $exec_string
```

```
name: "my_model"
  platform: "tensorrt_plan"
  max_batch_size: 64
  input [
    {
      name: "input_ids"
      data_type: TYPE_INT32
      dims: [ 768 ]
    },
    {
      name: "attention_mask"
      data_type: TYPE_BOOL
      dims: [ 768 , 768 ]
    },
    {
      name: "position_ids"
      data_type: TYPE_INT32
      dims: [ 768 ]
    }
  ]
  output [
    {
      name: "logits"
      data_type: TYPE_FP32
      dims: [ 256, 20 ]
    }
  ]
  instance_group {
    count: 2
    kind: KIND_GPU
}
dynamic_batching {
  preferred_batch_size: [64]
  max_queue_delay_microseconds: 0,
}
```
I benchmark using the same command (nvcr.io/nvidia/tritonserver:23.03-py3-sdk or nvcr.io/nvidia/tritonserver:23.04-py3-sdk didn't make difference)
```
perf_analyzer -m my_model  --concurrency-range=8:10  -b 16 --shape=input_ids:16,768 --shape=attention_mask:16,768,768 --shape=position_ids:16,768 --measurement-interval 15000 -u 0.0.0.0:8080
```

While benchmarking I can see that 23.03triton server is correctly generating dynamic batches of size 4(4*16=64) for each instance. While 23.04 isn't, despite trying multiple combinations of max_queue_delay_microseconds and perf_analyzer request concurrency & batches & model_analyzer.
Is there a major difference between the dynamic batch schedulers between the two versions?
The 23.04 result:
`Concurrency: 8, throughput: 28.1469 infer/sec, latency 4701133 usec` throughput moves between 26~28 
The 23.03 result:
`Concurrency: 8, throughput: 48.5902 infer/sec, latency 2657543 usec` throughput moves between 47~50 

Also weirdly tensorrt:23.04 is generating a much smaller model than tensorrt:23.03. I can even fit 3 instances vs 2.
Both of the models performed close to the original model, but haven't compared the outputs of both to each other.
`np.testing.assert_allclose(to_numpy(torch_output), pred, rtol=1e-04, atol=1e-04)`

Anyway this seems more like a dynamic batch scheduler problem than anything else.

**Triton Information**
Using the official containers.
NVIDIA-SMI 530.30.02   Driver Version: 530.30.02    CUDA Version: 12.1
Tesla T4

] [TRT] Input filename:   models/my_model.onnx
] [TRT] ONNX IR version:  0.0.7
] [TRT] Opset version:    14
] [TRT] Producer name:    pytorch
] [TRT] Producer version: 2.0.1
] [TRT] Domain:           
] [TRT] Model version:    0
] [TRT] Doc string:

**To Reproduce**
Although I can't share the model. It is an electra model architecture.

**Expected behavior**
There shouldn't be any difference due to dynamic batch scheduling.

**Comment by dyastremsky** [comment]
Thank you for reporting this potential bug. Are you seeing similar performance with concurrency == 1? We've filed a ticket to investigate. If you're able to provide a toy model that reproduces this issue, that'd be helpful for investigating.

**Comment by dyastremsky** [comment]
A couple of additional questions:
- Can you attach your full logs? I'd like to see the input shapes generated and what the formed batches are. I would think that using the batch flag would add an extra dimension of [16] to your inputs, but perhaps there is logic that knows not to do so.
- Please double check how you're generating your models. I'm not familiar with the specific updates from Tensor 23.03 vs 23.04, but if your model size is varying a lot, it'd be good to make sure they're identical.
- Have you tried running the TensorRT models outside of Triton, e.g. on trtexec? Do they get the same performance between 23.03 and 23.04.?

**Comment by cceyda** [comment]
> - Please double check how you're generating your models. I'm not familiar with the specific updates from Tensor 23.03 vs 23.04, but if your model size is varying a lot, it'd be good to make sure they're identical.

> - Have you tried running the TensorRT models outside of Triton, e.g. on trtexec? Do they get the same performance between 23.03 and 23.04.?

Yes my model is giving identical results for Tensor 23.03 vs 23.04. I have tested with trtexec and the performance is similar too. So I don't think it is due to tensorrt.

**Comment by dyastremsky** [comment]
Thanks for confirming. We'll look into this soon.

**Comment by GuanLuo** [comment]
Hi @cceyda, I wonder if you can try the latest release (23.06) and check if the regression is still present?

**Comment by GuanLuo** [comment]
Hi @cceyda, following up on whether you have tried the later release on performance regression. If it is still being observed on your side, do you mind to share your model for the reproducing the issue?

I have run with some models I have locally but I don't see a significant performance impact. The model is configured with the same batch size and instance setting in the issue description:
```
23.03
perf_analyzer -m resnet152_plan  --concurrency-range=8:10  -b 16  --measurement-interval 15000
Concurrency: 8, throughput: 1236.64 infer/sec, latency 103489 usec
Concurrency: 9, throughput: 1226.28 infer/sec, latency 117402 usec
Concurrency: 10, throughput: 1227.16 infer/sec, latency 130356 usec

perf_analyzer -m resnet152_plan  --concurrency-range=16:20  -b 16  --measurement-interval 15000
Concurrency: 16, throughput: 1809.16 infer/sec, latency 141269 usec
Concurrency: 17, throughput: 1826.22 infer/sec, latency 148940 usec
Concurrency: 18, throughput: 1798.93 infer/sec, latency 160032 usec
Concurrency: 19, throughput: 1808.44 infer/sec, latency 168190 usec
Concurrency: 20, throughput: 1820.54 infer/sec, latency 175638 usec

23.04
Concurrency: 8, throughput: 1345.08 infer/sec, latency 95163 usec
Concurrency: 9, throughput: 1697.65 infer/sec, latency 84812 usec
Concurrency: 10, throughput: 1540.91 infer/sec, latency 103807 usec

Concurrency: 16, throughput: 1852.54 infer/sec, latency 138019 usec
Concurrency: 17, throughput: 1926.24 infer/sec, latency 141119 usec
Concurrency: 18, throughput: 1807.86 infer/sec, latency 159237 usec
Concurrency: 19, throughput: 1938.8 infer/sec, latency 156884 usec
Concurrency: 20, throughput: 2063.79 infer/sec, latency 155047 usec

23.07
Concurrency: 8, throughput: 1737.9 infer/sec, latency 73639 usec
Concurrency: 9, throughput: 1675.73 infer/sec, latency 85926 usec
Concurrency: 10, throughput: 1740.32 infer/sec, latency 91928 usec

Concurrency: 16, throughput: 1793.8 infer/sec, latency 142530 usec
Concurrency: 17, throughput: 1773.19 infer/sec, latency 153275 usec
Concurrency: 18, throughput: 1861.94 infer/sec, latency 154747 usec
Concurrency: 19, throughput: 1931.42 infer/sec, latency 157367 usec
Concurrency: 20, throughput: 1970.43 infer/sec, latency 162267 usec
```

**Comment by cceyda** [comment]
I'll test on the next (23.07) release. I don't have the bandwidth to update my system right now.
Under real load (with a reasonable max_queue_delay_microseconds>100000 ) it performs similarly. Maybe using max_queue_delay_microseconds=0 with perf_analyzer is just unrealistic, although I like knowing the upperbound.
